Introduces PTP tab and Job

dbrembilla opened the pull request:
This MR introduces the first point to polygon functionalities, in particular:
- Introduces a Tab in the volume view for the conversion
![image](https://github.com/user-attachments/assets/4eb64a89-b15e-4500-8c33-01060834c33e)

- Creates a tab that can be used to create a PTP Job. 
![image](https://github.com/user-attachments/assets/3c8ebb9a-40e6-490c-9a61-6e394dc75304)

Once you have a Volume with point annotations, you can run a Point to Polygon job on a label with at least a point annotation. This will in turn run:
- A Job computing the Expected areas for the label
- A job uploading these expected areas
- A job running the conversion, using the median of these expected areas as the expected area for the job
- A job that uploads the new converted annotations

To test this MR:
- Use the `dev-modules-point-to-polygon` branch for the biigle root 
- Add PTP to the composer, for now using the VCS (If you think I should add it to Packagist let me know)
- Try to run the Point to Polygon conversion.

mzur commented on src/Http/Controllers/Views/PtpController.php:
```diff
@@ -0,0 +1,82 @@
+        $labels = Label::whereIn('label_tree_id', $labelTrees->pluck('id'))->get();
+
+        // ID of point shape
+        $pointAnnotationId = 1;
```
```suggestion
        $pointAnnotationId = Shape::pointId();
```

mzur commented on src/Jobs/PtpJob.php:
```diff
@@ -0,0 +1,236 @@
+use Illuminate\Queue\SerializesModels;
+use Storage;
+
+class PtpJob extends BaseJob implements ShouldQueue
```
This class has several attributes that were copy/pasted but are not relevant for this job.

mzur commented on src/Jobs/UploadConvertedAnnotationsJob.php:
```diff
@@ -0,0 +1,86 @@
+     */
+    public $queue;
+
+    /**
+     * User responsible for the Point to Polygon conversion.
+     *
+     * @var User
+     */
+    public User $user;
+
+    /**
+     * The queue to push this job to.
+     *
+     * @var string
+     */
+    public string $inputDir;
+
+
+    public function __construct(string $inputDir, User $user)
+    {
+        $this->inputDir = $inputDir;
+        $this->user = $user;
+    }
```
With new PHP syntax you can do:

```suggestion
    /**
     * @var $user User responsible for the Point to Polygon conversion.
     * @var $inputDir ...
     */
    public function __construct(public string $inputDir, public User $user)
    {
       //
    }
```

mzur commented on src/Jobs/UploadConvertedAnnotationsJob.php:
```diff
@@ -0,0 +1,86 @@
+            if (is_null($jsonData)) {
+                throw new Exception("Error while reading file $file");
+            }
+            $polygonShape = 3;
```
```suggestion
            $polygonShape = Shape::polygonId();
```

mzur commented on src/Jobs/PtpJob.php:
```diff
@@ -0,0 +1,236 @@
+        $this->jobType = $jobType;
+        foreach ($annotations as $annotation) {
+            $this->points[] = implode(',', $annotation['points']);
+            //TODO: find a better way to pass this arg.
+            $this->annotationIds[] = $annotation['annotation_id'];
```
In the past when I had too many input parameters to a Python script I would write everything into a JSON file and pass this to the Python script.

With the annotation IDs as inline parameter you could easily run into the CLI parameter limit.

mzur commented on src/Jobs/PtpJob.php:
```diff
@@ -0,0 +1,236 @@
+            $medianArea = $this->findMedian($expectedAreaValues);
+            $command = $command." -e $medianArea";
+        }
+        exec("$command > {$logFile} 2>&1", $lines, $code);
+
+
+        if ($code !== 0) {
+            $lines = File::get($logFile);
+            throw new Exception("Error while executing python script'{$script}':\n{$lines}", $code);
+        }
```
There should be no need to specify an extra log file ([example](https://github.com/biigle/magic-sam/blob/925439991751d088ab0b46a543d49772442884ac/src/Jobs/GenerateEmbedding.php#L133-L138)).

mzur commented on src/resources/scripts/ptp.py:
```diff
@@ -0,0 +1,835 @@
+    mask = mask * 255
+    mask = mask.astype(np.uint8)
+    contours, hierarchy = cv2.findContours(
+        mask, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_NONE
```
We don't want to store coordinates for each pixel :wink: 

```suggestion
        mask, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE
```

mzur commented on src/resources/scripts/ptp.py:
I haven't read the whole script but maybe you can boil it down to only those features that are actually used by this BIIGLE module.

mzur commented on src/resources/views/volumesSidebar.blade.php:
```diff
@@ -1,3 +1,3 @@
 @if (($user->can('edit-in', $volume) || $user->can('sudo')) && $volume->isImageVolume() && !$volume->hasTiledImages())
-    <sidebar-tab name="maia" icon="robot" title="Perform Machine Learning Assisted Image Annotation (MAIA)" href="{{route('volumes-maia', $volume->id)}}"></sidebar-tab>
+    <sidebar-tab name="ptp" icon="circle" title="Perform Point to Polygon conversion" href="{{route('volumes-ptp-conversion', $volume->id)}}"></sidebar-tab>
```
I thought that maybe we could use the same "magic hat" icon than MagicSAM, as the feature is very similar. It's basically MagicSAM on a volume level. @dlangenk what do you think?

dlangenk replied:
Yeah it would be nice to use the same logo.

mzur commented on src/Http/Controllers/Api/PtpController.php:
```diff
@@ -0,0 +1,91 @@
+     */
+    public function generatePtpJob(Request $request) {
+        $this->validate($request, ['label_id' => 'integer', 'volume_id' => 'integer']);
+        $volume = Volume::findOrFail($request->volume_id);
```
You have to validate that the volume is an image volume and (maybe) that is does not have tiled images here. Video volumes are not supported.

@dlangenk Would the Python script work with a huge image mosaic?

dlangenk replied:
I haven't tried it and currently it wouldn't work on tiles. But if you would get the full mosaic and use the right image loading library I don't see anything that would speak against that.

dbrembilla replied:
Would it be better to disallow computation of volumes with tiled images completely or to compute it with all images that are not tiled?

mzur commented on src/Http/Controllers/Views/PtpController.php:
```diff
@@ -0,0 +1,82 @@
+        // ID of point shape
+        $pointAnnotationId = 1;
+
+        //annotation with labels and that are points
+        $annotations = ImageAnnotation::join('image_annotation_labels','image_annotations.id', '=', 'image_annotation_labels.annotation_id')
+            ->join('images','image_annotations.image_id','=','images.id')
+            ->join('labels', 'image_annotation_labels.label_id','=','labels.id')
+            ->where('images.volume_id', $volume->id)
+            ->whereIn('image_annotation_labels.label_id', $labels->pluck('id'))
+            ->where('image_annotations.shape_id', $pointAnnotationId)
+            ->select('images.uuid', 'image_annotations.id', 'image_annotation_labels.label_id', 'labels.name AS label_name')->get();
+
+        //largo URL patches to show
+        $largoPatchesUrl = Storage::disk(config('largo.patch_storage_disk'))->url(':prefix/:id.'.config('largo.patch_format'));
```
These are not needed (any more).

mzur reviewed (changes requested):
Thanks! Mostly high-level comments first:

**GPU or CPU?**

Do you think these jobs should run on a GPU or is the runtime also decent on a CPU? We have to consider the following options:

1. On the GPU the whole volume is processed in a single job (because we have only one GPU for these long jobs). This may have the advantage of having to cache all files only once. Also, we may not need a temporary storage for the expected areas. The Python script can just get an input JSON with all annotation coordinates and label IDs as well as a directory of images and do everything else internally. The result is a JSON with the converted polygon coordinates.

2. On the CPU we can parallelize things and compute each image in a single job. This needs the temporary storage for the expected area and the different types of jobs (first process images for expected area, then do conversion).

Your current implementation of the jobs targets 2. Maybe it will be faster and easier to implement to do this on a GPU?

**Volume Tab**

I'd like this iteration of the feature to be implemented as an actual tab in the volume overview instead of a completely new view. You can take the biigle/laserpoints tab as an example.

Selection of a subset of labels should be optional. Either we don't offer this option at all for now (i.e. the user can only click the button "start" and all point annotations are converted) or we allow a selection of multiple labels. This can use the existing component to select multiple labels from label trees. Example from the clone volumes view:

![Screenshot from 2024-12-06 14-15-41](https://github.com/user-attachments/assets/2c53c645-57cb-4b1d-9d03-5d1db5c81fbd)

**Tests**

You have to write tests for the controller and job classes.

**Garbage files**

There are still some files from biigle/maia in the repo (Requests, Jobs). Also the console publish command is not needed (it was used in the past by modules but not any more).

**Branch**

There is already the `dev-modules-ptp` branch of biigle/core with the module installed. However, the module is curently not loaded there. You should add an [extra section](https://github.com/biigle/magic-sam/blob/925439991751d088ab0b46a543d49772442884ac/composer.json#L22-L28) to the composer.json to enable package discovery upon installation.

mzur commented:
@dbrembilla You requested a review but then added new commits after that. Is this ready now?

dbrembilla commented:
> @dbrembilla You requested a review but then added new commits after that. Is this ready now?

Sorry @mzur, I realized I missed some things. Now it should be good to go!

mzur commented on src/Http/routes.php:
```diff
@@ -2,49 +2,14 @@
-
-    $router->get('volumes/filter/knowledge-transfer', 'KnowledgeTransferVolumeController@index');
-    $router->get('volumes/filter/area-knowledge-transfer', 'AreaKnowledgeTransferVolumeController@index');
+    $router->get('ptp/{id}', ['as' => 'volumes-ptp-conversion', 'uses' => 'PtpController@index']);
```
No longer required?

mzur commented on src/PtpServiceProvider.php:
```diff
@@ -36,6 +36,7 @@ public function boot(Modules $modules, Router $router)
         $modules->register('ptp', [
             'viewMixins' => [
                 'volumesSidebar',
+                'ptpContainer',
```
This does not exist.

mzur commented on tests/Http/Controller/PtpControllerTest.php:
```diff
@@ -0,0 +1,46 @@
+
+class PtpControllerTest extends ApiTestCase
+{
+    public function testCreateJob()
+    {
+        Queue::fake();
+        $image = Image::factory()->create(['volume_id' => $this->volume()->id]);
+
+        config(['ptp.ptp_storage_disk' => 'test']);
+
+        $this->doTestApiRoute('POST', '/api/v1/send-ptp-job');
+
+        $this->beGlobalGuest();
+        $this->postJson('/api/v1/send-ptp-job', ['volume_id' => $this->volume()->id])->assertStatus(403);
+
+        $this->beUser();
+        $this->postJson('/api/v1/send-ptp-job', ['volume_id' => $this->volume()->id])->assertStatus(403);
+
+        $this->beEditor();
+        $this->postJson('/api/v1/send-ptp-job', ['volume_id' => $this->volume()->id])
+            ->assertStatus(200);
+
+        $v = VolumeTest::create(['media_type_id' => MediaType::videoId()]);
+
+        $this->beEditor();
+        $this->postJson('/api/v1/send-ptp-job', ['volume_id' => $v->id])->assertStatus(400);
+
+        $v2 = VolumeTest::create(['media_type_id' => MediaType::imageId()]);
+        $image = Image::factory()->create(['volume_id' => $v2->id, 'tiled' => true]);
+
+        $this->beEditor();
+        $this->postJson('/api/v1/send-ptp-job', ['volume_id' => $v2->id])->assertStatus(400);
+     }
```
Please separate these into multiple test methods/cases. You have one default case where you check the API route and authorization. Then you have another case where you check video volumes. Finally a third case with tiled images. The method names should reflect this so you can roughly know what is wrong when a test case failed.

dbrembilla commented on src/resources/scripts/ptp.py:
The script should now contain only code that is pertinent to the point to polygon conversion. It will attempt to use multiple prompt strategies so that if one does not produce a satisfying result, it will attempt other ways

mzur commented on src/Database/Factories/PtpExpectedAreaFactory.php:
I'm all for keeping it simple at the beginning and adding complexity later :+1:

mzur commented on tests/Jobs/PtpJobTest.php:
```diff
@@ -0,0 +1,116 @@
+        $this->volume = Volume::factory()->create();
+        $this->image = Image::factory()->create(['volume_id' => $this->volume->id]);
+        $attrs = $this->volume->attrs;
+        $this->uuid = Uuid::uuid4();
+        $attrs['ptp_job_id'] = $this->uuid;
+        $this->volume->attrs = $attrs;
+        $this->volume->save();
+        $this->imageAnnotation = ImageAnnotation::factory()->create([
+            'image_id' => $this->image->id,
+        ]);
+        $this->label = Label::factory()->create();
+        $this->user = User::factory()->create();
+        $this->user2 = User::factory()->create();
+        $this->imageAnnotationLabel = ImageAnnotationLabel::factory()->create([
+            'annotation_id' => $this->imageAnnotation->id,
+            'label_id' => $this->label->id,
+            'user_id' => $this->user->id,
+        ]);
+        $this->inputFile = 'test';
+        $inputFileContent = [
+            $this->image->id => [[
+                'annotation_id' => $this->imageAnnotation->id,
+                'points' => [1,2],
+                'shape' => Shape::pointId(),
+                'image' => $this->image->id,
+                'label' => $this->label->id,
+            ]]];
+        $this->outputFile = 'testOutput.json';
+        $outputFileContent = [[
+            'points' => [1,2,3,4,5,6],
+            'annotation_id' => $this->imageAnnotation->id,
+            'label_id' => $this->label->id,
+        ]];
+
+        config(['ptp.ptp_storage_disk' => 'test']);
+        config(['ptp.device' => 'cpu']);
+        $storage = Storage::disk('test');
+        $storage->put(config('ptp.temp_dir').'/'.$this->inputFile.'.json', json_encode($inputFileContent));
+        file_put_contents(config('ptp.temp_dir').'/'.$this->outputFile, json_encode($outputFileContent));
+    }
```
Some of this is not required in all of the tests (as far as I can see). Particularly creating files on disk should be avoided wherever possible as it makes the tests slow. But *when* you create files, you also have to take care of cleaning them up after the test or when the test fails. I usually do it like this:

```php
// create file
try {
   // test stuff
} finally {
   // delete file
}
```

mzur reviewed (changes requested):
This is progressing nicely :+1:

dbrembilla commented on src/Jobs/PtpJob.php:
```diff
@@ -0,0 +1,180 @@
+            throw new Exception("Error while reading file $this->outputFile");
+        }
+        $polygonShape = Shape::polygonId();
+        foreach ($jsonData as $annotation) {
+            $newAnnotation = ImageAnnotation::findOrFail($annotation['annotation_id'])->replicate();
+            $newAnnotation->points = $annotation['points'];
+            $newAnnotation->shape_id = $polygonShape;
+            $newAnnotation->save();
+            $imageAnnotationLabel = [
+                'label_id' => $annotation['label_id'],
+                'annotation_id' => $newAnnotation->id,
+                'user_id' => $this->user->id,
+                'confidence' => 1.0,
+            ];
+            ImageAnnotationLabel::insert($imageAnnotationLabel) ;
+        }
```
Is it okay if I literally copy paste it with the objective later to create a trait for it?

dbrembilla commented on src/config/ptp.php:
```diff
@@ -29,18 +29,44 @@
+    |
+    | Devices: cpu, cuda
+    */
+    'device' => env('PTP_DEVICE', 'cpu'),
```
How do you do that?  Is it sufficient that it is forced to use the gpu queue?

dbrembilla commented:
@mzur I set up the notifications and added new tests. To test that if the PtpJob failes the construct method I added a test condition that makes it throw an Exception. I was looking for a way to monkeypatch the Job, but apparently I would need to add a new dependency. Do you think this is good or should I remove it?

mzur commented on tests/Jobs/PtpJobTest.php:
```diff
@@ -21,94 +21,182 @@ class PtpJobTest extends TestCase
+
+        $this->expectException(Exception::class);
+        $this->expectExceptionMessage('No files converted!');
+        try {
+            file_put_contents(config('ptp.temp_dir').'/'.$this->outputFile, '[]');
+            $this->setUpAnnotations();
+            $job->handle();
+            $this->assertTrue($job->pythonCalled);
+            $volume = Volume::where('id', $this->volume->id)->first();
+            $this->assertFalse(isset($volume->attrs['ptp_job_id']));
+        } catch(Exception $e) {
+            throw ($e);
+        } finally {
+            unlink(config('ptp.temp_dir').'/'.$this->inputFile.'.json');
+            unlink(config('ptp.temp_dir').'/'.$this->inputFile.'_images.json');
+        }
```
I said above that you don't need the input/output paths as constructor parameter any more but here you use them for testing. What you con do instead is to update the mock class below so you can set something like a public `inputFileContent` array. Also you can update the mocked `python()` method to write something to a public `outputFileContent` variable in the mock class.

This way you avoid test code in the actual implementation.

mzur commented on tests/Jobs/PtpJobTest.php:
```diff
@@ -21,94 +21,182 @@ class PtpJobTest extends TestCase
+        config(['ptp.python' => 'fake']);
+        try {
+            $job->handle();
+        } catch(Exception $e) {
+            throw ($e);
+        } finally {
```
No need for this
```suggestion
        } finally {
```

mzur commented on tests/Jobs/PtpJobTest.php:
```diff
@@ -21,94 +21,182 @@ class PtpJobTest extends TestCase
+            $this->assertTrue($job->pythonCalled);
+            $volume = Volume::where('id', $this->volume->id)->first();
+            $this->assertFalse(isset($volume->attrs['ptp_job_id']));
+        } catch(Exception $e) {
+            throw ($e);
+        } finally {
```
```suggestion
        } finally {
```

mzur commented on src/Http/Controllers/Api/PtpController.php:
```diff
@@ -0,0 +1,89 @@
+        }
+
+        $pointShapeId = Shape::pointId();
+        $annotationsCount = ImageAnnotation::join('images','image_annotations.image_id', '=','images.id')
+            ->where('images.volume_id', $volumeId)
+            ->where('image_annotations.shape_id', $pointShapeId)
+            ->count();
+
+        if ($annotationsCount == 0){
```
Better use `exists()` instead of `count()`.

mzur commented on src/Jobs/PtpJob.php:
```diff
@@ -0,0 +1,324 @@
+{
+    use Batchable, InteractsWithQueue, Queueable, SerializesModels;
+
+    /**
+     *
+     * @var string $tmpInputFile File where the input data for the Python script will be kept
+     * @var string $tmpImageInputFile File where the image input data for the Python script will be kept
+     * @var int $insertChunkSize Number of annotations to be inserted per chunk
+     *
+     */
+    private string $tmpInputFile;
+    private string $tmpImageInputFile;
+    public static int $insertChunkSize = 5000;
```
Please use one docblock for each variable.

mzur commented on src/Jobs/PtpJob.php:
```diff
@@ -0,0 +1,324 @@
+
+        $pointShapeId = Shape::pointId(); //Find annotations with selected label in desired volume
+        $annotations = ImageAnnotation::join('image_annotation_labels','image_annotations.id', '=', 'image_annotation_labels.annotation_id')
+            ->join('images','image_annotations.image_id', '=','images.id')
```
```suggestion
            ->join('images','image_annotations.image_id', '=', 'images.id')
```

mzur commented on src/Jobs/PtpJob.php:
```diff
@@ -0,0 +1,324 @@
+    {
+        $imageAnnotationArray = [];
+
+        $pointShapeId = Shape::pointId(); //Find annotations with selected label in desired volume
```
This comment does not describe the current line. Generally, describe in a comment *why* something is done instead of *what* (because *what* can be determined from the code). Of course there are exceptions when the code is very complicated. I don't think a comment is needed here.

```suggestion
        $pointShapeId = Shape::pointId();
```

mzur commented on src/Jobs/PtpJob.php:
```diff
@@ -0,0 +1,324 @@
+            ];
+        };
+
+        //$inputFile.'.json' will be used for image annotations, $inputFile.'_images.json' for image paths
```
I feel this comment belongs elsewhere because this method has nothing to do with the "images" input file.

mzur commented on src/Jobs/PtpJob.php:
```diff
@@ -0,0 +1,324 @@
+        $jsonData = json_encode($imageAnnotationArray);
+
+        //Create input file with annotations
+        if (file_exists($this->tmpInputFile)) {
+            unlink($this->tmpInputFile);
```
Shouldn't this be an error condition? Each job should use a unique file. In any case, `file_put_contents` overwrites files by default, so no deletion would be necessary.

dbrembilla replied:
The case in which there is already a file here is if there was some problem here when running the script, if fails after having created  the file and we start the program again and the input file is already present. 

I'd be more inclined to not raising an error here. it would also be consistent I believe, since if it runs on different workers the collision would not happen as they do not share the `tmp` folder (correct me if I'm wrong, maybe they use the same system tmp somehow?). 

If `file_put_contents` normally overwrites the file, I guess that would be okay.

mzur commented on src/Jobs/PtpJob.php:
```diff
@@ -0,0 +1,324 @@
+        if (file_exists($this->tmpInputFile)) {
+            unlink($this->tmpInputFile);
+        } else if (!file_exists(dirname($this->tmpInputFile))){
+            mkdir(dirname($this->tmpInputFile), recursive:true);
```
```suggestion
            mkdir(dirname($this->tmpInputFile), recursive: true);
```

mzur commented on src/Jobs/PtpJob.php:
```diff
@@ -0,0 +1,324 @@
+        $this->maybeDownloadCheckpoint($checkpointUrl, $modelPath);
+
+        if (!file_exists(dirname($this->outputFile))) {
+            mkdir(dirname($this->outputFile), recursive:true);
```
```suggestion
            mkdir(dirname($this->outputFile), recursive: true);
```

mzur commented on src/Jobs/PtpJob.php:
```diff
@@ -0,0 +1,324 @@
+
+        if (!file_exists(dirname($this->outputFile))) {
+            mkdir(dirname($this->outputFile), recursive:true);
+        } else if (file_exists($this->outputFile)){
+            unlink($this->outputFile);
+        }
```
Same here, files should not collide.

mzur commented on src/Jobs/PtpJob.php:
```diff
@@ -0,0 +1,324 @@
+    /**
+     * Upload the converted annotations to the DB
+     *
+    **/
```
```suggestion
     */
```

mzur commented on src/Jobs/PtpJob.php:
```diff
@@ -0,0 +1,324 @@
+            $insertAnnotationLabels[] = [
+                'label_id' => $annotation['label_id'],
+                'user_id' => $this->user->id,
+                'confidence' => 1.0,
```
The confidence will be set in `insertAnnotationChunk`.

```suggestion
```

mzur commented on src/Notifications/PtpJobConcluded.php:
```diff
@@ -0,0 +1,80 @@
+     * @param Report $report
+     * @return void
+     */
+    public function __construct(string $volume)
```
This is the volume *name* only, right? If you want to keep it like that, this should be reflected in the variable name and comments.

You could also pass the entire volume, too, which would allow you to add a handy button to open the volume to the notification ([example](https://github.com/biigle/maia/blob/ab9dd4c3a61d81b734b87d42c2311b06839d33d1/src/Notifications/JobStateChanged.php#L63-L65)).

mzur commented on src/Notifications/PtpJobFailed.php:
```diff
@@ -0,0 +1,80 @@
+     * @param Report $report
+     * @return void
+     */
+    public function __construct(string $volume)
```
See above.

mzur commented on src/config/ptp.php:
```diff
@@ -29,22 +29,49 @@
+    'ptp_script' => __DIR__.'/../resources/scripts/ptp.py',
 
     /*
-    | URL from which to download the trained weights for the model.
+    | Path to the script that performs PTP conversion.
+    */
+    'ptp_storage_disk' => env('PTP_STORAGE_DISK', 'ptp'),
```
This should no longer be needed.

mzur commented on src/resources/assets/js/ptpContainer.vue:
```diff
@@ -0,0 +1,57 @@
+        },
+
+        sendPtpRequest() {
+            PtpJobApi.save({id: this.volumeId}, {})
```
```suggestion
            PtpJobApi.save({id: this.volumeId})
```

dbrembilla replied:
So, if I do this, for some reason it won't add the variables to the API route (e.g. it sends a request to `api/v1/send-ptp-job` rather than `api/v1/send-ptp-job/1`.) That is why I put the `, {}` after it. While I did not find any documentation about it, [here](https://stackoverflow.com/questions/37045717/vue-js-vue-resource-calling-resource-save-with-path-parameter) it is suggested that if only one parameter is added, it is handled as the request payload. Do you have some input about this?

mzur commented on src/resources/assets/js/ptpContainer.vue:
```diff
@@ -0,0 +1,57 @@
+              <button
+                  class="btn btn-success btn-block"
+                  type="button"
+                  title="Run Point to Polygon conversion on this volume"
```
If the button is disabled because a job is already running, the title should change to something that explains this.

mzur commented on src/resources/views/volumesSidebar.blade.php:
```diff
@@ -1,3 +1,12 @@
 @if (($user->can('edit-in', $volume) || $user->can('sudo')) && $volume->isImageVolume() && !$volume->hasTiledImages())
```
Global admins actually can't create annotations in projects they don't belong to (and the `edit-in` authorization check will fail), so the tab can be hidden for them, too.

```suggestion
@if ($user->can('edit-in', $volume) && $volume->isImageVolume() && !$volume->hasTiledImages())
```

mzur commented on tests/Http/Controller/PtpControllerTest.php:
```diff
@@ -0,0 +1,147 @@
+            'shape_id' => Shape::pointId(),
+        ]);
+        $this->beEditor();
+        config(['ptp.ptp_fail_construct' => true]);
+        config(['ptp.ptp_storage_disk' => 'test']);
```
Instead of the test config, you can probably use something like `Bus::shouldReceive(...)->andThrow(\Exception:.class)` (see [Laravel docs](https://laravel.com/docs/11.x/queues#testing)).

mzur reviewed (changes requested):
Adding test code to the actual class is not a good practice. Use the `MockPtpJob` that you already define in `PtpJobTest.php` instead. You can override the original constructor there:

```php
public function __construct(
   public int $volumeId,
   public string $volumeName,
   public string $inputFile,
   public string $outputFile,
   public User $user,
   public string $id,
   bool $fail = false,
)
{
   if ($fail) {
      throw new Exception("Error!");
   }

   $args = array_slice(func_get_args(), 0, -1, true);

   parent::__construct(...$args);
}
```

Other comments:

- Please care more about coding style 😉

- Consider to use the `File` facade instead of native methods such as `unlink` or `file_put_contents`. This can sometimes help in tests. Also it will be consistent with the rest of the codebase.

More comments below.

dbrembilla commented on src/Http/Controllers/Api/PtpController.php:
```diff
@@ -6,79 +6,74 @@
+        try {
+            PtpJob::dispatch($volume->id, $volume->name, $inputFile, $outputFile, $request->user(), $id);
+        } catch (Exception $e) {
+            // If unable to dispatch a PTP Job, reset the PTP Job ID
+            $attrs = $volume->attrs;
+            unset($attrs['ptp_job_id']);
+            $volume->attrs = $attrs;
+            $volume->save();
+            abort(400, 'Unable to create Point To Polygon conversion job!');
```
You mean let it fail hard without resetting the volume status?

dbrembilla commented on src/Jobs/PtpJob.php:
```diff
@@ -46,68 +78,109 @@ public function __construct(public string $inputFile, public string $outputFile,
-        $imageIds = array_keys($storage->json($this->inputFile.'.json'));
+        $imageData = $this->generateInputFile();
+        $imageIds = array_keys($imageData);
         $images = Image::whereIn('id', $imageIds)->get()->all();
```
Yes, I have the image IDs, but since the `File::Batch` function requires that the first element is an array of `Image` object, I guess  I still have to get them somewhere. A solution I found  was to get them as the annotations are gatherered in the `generateInputFile` function, and adding a `with('file')` statement to the query there  which now uses `lazy()`, as you suggested below. Only the `Image` would be returned by this function. This should be better if I understand correctly. But maybe it'd  be better to show you once I send the request for review again I guess.

mzur commented on tests/Http/Controller/PtpControllerTest.php:
```diff
@@ -113,35 +109,7 @@ public function testNoImageAnnotations()
-        $volume = Volume::where('id', $this->volume()->id)->first();
-
-        $this->assertFalse(isset($volume->attrs['ptp_job_id']));
-    }
```
This must be reverted, I'm afraid. Last time I suggested to use `Bus::shouldReceive(...)->andThrow(\Exception::class)` instead of a test config variable.

mzur reviewed (changes requested):
Sorry for the long wait!

Some of the comments of the previous review are still open. You'll see my replies below but the conversations are still above.

I saw that you already updated the JS code for Vue 3, nice! You can remove the `.npmrc` and `webpack.mix.js` files then.

Please add a manual article that explains what this module does and link it directly from the PTP tab. You can add a button/link to the top right similar to the laserpoints tab. Also I think we sould use the tab title "convert points to polygons". Since the tab is mostly empty, you could even add a very short paragraph explaining what it does there.

The job queue and connection can be configured in the config but the config is never used. Since this is supposed to be a GPU job, we have to be able to configure these.

When I run the PTP job (it works on a CPU if slowly), I get the following error:

```
Traceback (most recent call last):
  File "/var/www/vendor/biigle/ptp/src/config/../resources/scripts/ptp.py", line 777, in <module>
    expected_areas.groupby("label_id")
  File "/opt/conda/lib/python3.11/site-packages/pandas/core/frame.py", line 9183, in groupby
    return DataFrameGroupBy(
           ^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/pandas/core/groupby/groupby.py", line 1329, in __init__
    grouper, exclusions, obj = get_grouper(
                               ^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/pandas/core/groupby/grouper.py", line 1043, in get_grouper
    raise KeyError(gpr)
KeyError: 'label_id'
```

This was the content of the input file:

```json
{"1":[{"annotation_id":1,"points":[1872.56,1201.11],"shape":1,"image":1,"label":2}],"2":[{"annotation_id":6,"points":[2139.64,1188],"shape":1,"image":2,"label":1}]}
```

More comments below.

dbrembilla commented:
@mzur I guess the conversion failed because it was unable to generate any expected area. I added a check in the python script.

dbrembilla commented on src/resources/scripts/ptp.py:
```diff
@@ -771,6 +771,10 @@ def process_image(
 
+    if len(resulting_annotations) == 0:
+        logging.error("Unable to compute any annotations for expected area!")
+        exit(0)
```
Here I made the Python script do `exit(0)` if no expected areas were computed - I wanted to avoid the entire job breaking if only one image had a problem in the conversion.

dbrembilla commented:
Hi, sorry for the long wait.
1. I added and tested the addition of the `ProcessAnnotatedJobs` after the generation of the new annotations
2. Yup, that works like that, so the person who starts the PTP job will be the author of the new annotations. See line 255 in `PtpJobs.php`
3. Right now, if there are multiple labels attached to an annotations, they are counted multiple times and can result in multiple annotations. Do you think it should work as a single annotation with multiple labels? In my opinion for now it should work like this, so that if the labels have different expected areas this will be respected, and if the behavior should be that the annotations should still be the same this should be done manually. But I also see that it should work as a single annotation with multiple labels. It could also potentially be something we could fix as we see the opinion of the users and so. But I'd like to hear your opinion about this.
